go/tendermint/apps/staking/state/gas.go

```diff
@@ -0,0 +1,92 @@
+package state
+
+import (
+	"fmt"
+
+	"github.com/oasislabs/oasis-core/go/common/consensus/gas"
+	"github.com/oasislabs/oasis-core/go/common/crypto/signature"
+	"github.com/oasislabs/oasis-core/go/common/quantity"
+	staking "github.com/oasislabs/oasis-core/go/staking/api"
+	"github.com/oasislabs/oasis-core/go/tendermint/abci"
+)
+
+// feeAccumulatorKey is the block context key.
+type feeAccumulatorKey struct{}
+
+func (fak feeAccumulatorKey) NewDefault() interface{} {
+	return &feeAccumulator{}
+}
+
+// feeAccumulator is the per-block fee accumulator that gets all fees paid
+// in a block.
+type feeAccumulator struct {
+	balance quantity.Quantity
+}
+
+// AuthenticateAndPayFees authenticates the message signer and makes sure that
+// any gas fees are paid.
+//
+// This method modifies the common pool balance to transfer collected gas fees.
+//
+// When executed in a CheckTx context, this method returns a nil account as
+// the caller should not do any further processing.
+func AuthenticateAndPayFees(
+	ctx *abci.Context,
+	state *MutableState,
+	id signature.PublicKey,
+	nonce uint64,
+	fee *gas.Fee,
+) (*staking.Account, error) {
+	// Fetch account and make sure the nonce is correct.
+	account := state.Account(id)
+	if account.General.Nonce != nonce {
+		logger.Error("invalid account nonce",
+			"account_id", id,
+			"account_nonce", account.General.Nonce,
+			"nonce", nonce,
+		)
+		return nil, staking.ErrInvalidNonce
+	}
+
+	if ctx.IsCheckOnly() {
```

Comment:
Should the balance check be executed unconditionally?

Reply:
Well for the non-check case the balance check happens in `Move`. I will add a comment that this is the case.